runtime/runtime/src/congestion_control.rs

```diff
@@ -422,23 +477,34 @@ impl DelayedReceiptQueueWrapper {
         receipt: &Receipt,
         config: &RuntimeConfig,
     ) -> Result<(), RuntimeError> {
-        let delayed_gas = receipt_congestion_gas(receipt, &config)?;
-        let delayed_bytes = receipt_size(receipt)? as u64;
-        self.new_delayed_gas = safe_add_gas(self.new_delayed_gas, delayed_gas)?;
-        self.new_delayed_bytes = safe_add_gas(self.new_delayed_bytes, delayed_bytes)?;
-        self.queue.push(trie_update, receipt)?;
+        let gas = compute_receipt_congestion_gas(&receipt, &config)?;
+        let size = compute_receipt_congestion_size(&receipt)? as u64;
+
+        let receipt = match config.use_state_stored_receipt {
+            true => {
+                let metadata =
+                    StateStoredReceiptMetadata { congestion_gas: gas, congestion_size: size };
+                let receipt = StateStoredReceipt::new_borrowed(receipt, metadata);
+                ReceiptOrStateStoredReceipt::StateStoredReceipt(receipt)
+            }
+            false => ReceiptOrStateStoredReceipt::Receipt(Cow::Borrowed(receipt)),
```

Comment:
Now that I think about it, one could also make an argument that it's better for `push` to consume the receipt, not borrow it. It'd ensure that a receipt is never saved twice by mistake. :thinking:

Reply:
I attempted that actually. If I recall correctly in one of the usages (incoming
receipts perhaps) I couldn't get the ownership because the receipt was still
needed for the balance check at the end. It's probably doable but would require
some more refactoring. Let me add a todo as I think it was the only reason for
Cow and getting rid of that would be nice.